File: market-rotation-tracker/src/services/stockScorer.ts
import { DailyBar, SectorScore, StockSignal } from '../types';
import { scorePressurePoints } from './pressurePoints';
import { STOCK_WATCHLIST } from '../config/stockWatchlist';

// Score individual stocks for the top leader sectors.
// Uses parent sector rotation strength as macro/rotation proxy,
// then pressure points and RSI for timing.
export function scoreStockWatchlist(
    leaders: SectorScore[],
    barsBySymbol: Map<string, DailyBar[]>
): StockSignal[] {
    const signals: StockSignal[] = [];
    const seen = new Set<string>(); // deduplicate (e.g. NVDA in XLK + SMH)

    for (const sector of leaders) {
        const stocks = STOCK_WATCHLIST[sector.symbol];
        if (!stocks) continue;

        for (const stock of stocks) {
            if (seen.has(stock.symbol)) continue;
            seen.add(stock.symbol);

            const bars = barsBySymbol.get(stock.symbol);
            if (!bars || bars.length < 20) continue;

            const pressure = scorePressurePoints(stock.symbol, bars);

            // ── Score components ─────────────────────────────────────────────
            let score = 0;

            // Parent sector rotation strength (40%)
            // Inherit from the sector's RS — the sector is already a leader
            const rs = sector.relativeStrength;
            score += rs > 3 ? 40 : rs > 1 ? 30 : 20;

            // Pressure point quality (40%) — strong signals are high-value setups
            const strongPts  = pressure.points.filter(p => p.strength === 'strong').length;
            const moderatePts = pressure.points.filter(p => p.strength === 'moderate').length;
            score += Math.min(40, strongPts * 15 + moderatePts * 8);

            // RSI timing (20%) — favor neutral/mild RSI for clean entries
            const rsi = pressure.rsi;
            if (rsi >= 45 && rsi <= 65)      score += 20; // ideal entry zone
            else if (rsi > 65 && rsi < 72)   score += 10; // mildly extended, still ok
            else if (rsi >= 30 && rsi < 45)  score += 14; // mild pullback in uptrend
            else if (rsi < 30)               score += 16; // oversold — potential bounce
            else                             score += 3;  // overbought (>72) — avoid

            score = Math.min(100, Math.round(score));

            const conviction =
                score >= 70 ? 'high'   :
                score >= 50 ? 'medium' :
                score >= 30 ? 'low'    :
                'noise';

            const pressureSummary = pressure.points.length > 0
                ? pressure.points.map(p => p.description).join('; ')
                : 'No significant pressure points';

            const reasoning = [
                `Parent: ${sector.label} (RS ${rs > 0 ? '+' : ''}${rs}% vs SPY, ${sector.trend})`,
                `Pressure: ${pressureSummary}`,
                `RSI: ${rsi}`,
            ].join(' | ');

            signals.push({
                symbol: stock.symbol,
                label: stock.label,
                parentSector: sector.symbol,
                parentSectorLabel: sector.label,
                parentRotationScore: sector.score,
                pressurePoints: pressure.points,
                rsi,
                score,
                conviction,
                reasoning,
            });
        }
    }

    return signals
        .filter(s => s.conviction !== 'noise')
        .sort((a, b) => b.score - a.score);
}

File: market-rotation-tracker-kotlin/src/main/kotlin/com/rotationtracker/services/StockScorer.kt
package com.rotationtracker.services

import com.rotationtracker.config.STOCK_WATCHLIST
import com.rotationtracker.models.DailyBar
import com.rotationtracker.models.SectorScore
import com.rotationtracker.models.ShortInterestRecord
import com.rotationtracker.models.StockSignal

fun scoreStockWatchlist(
    leaders: List<SectorScore>,
    barsBySymbol: Map<String, List<DailyBar>>,
    shortInterestBySymbol: Map<String, ShortInterestRecord> = emptyMap(),
): List<StockSignal> {
    val signals = mutableListOf<StockSignal>()
    val seen    = mutableSetOf<String>() // deduplicate (e.g. NVDA in XLK + SMH)

    for (sector in leaders) {
        val stocks = STOCK_WATCHLIST[sector.symbol] ?: continue

        for ((symbol, label) in stocks) {
            if (!seen.add(symbol)) continue

            val bars = barsBySymbol[symbol] ?: continue
            if (bars.size < 20) continue

            val pressure = scorePressurePoints(symbol, bars, shortInterestBySymbol[symbol])
            var score    = 0

            // Parent sector rotation strength (40%)
            score += when {
                sector.relativeStrength > 3 -> 40
                sector.relativeStrength > 1 -> 30
                else -> 20
            }

            // Pressure point quality (40%)
            val strongPts    = pressure.points.count { it.strength == "strong" }
            val moderatePts  = pressure.points.count { it.strength == "moderate" }
            score += minOf(40, strongPts * 15 + moderatePts * 8)

            // RSI timing (20%) — favour neutral/mild RSI for clean entries
            val rsi = pressure.rsi
            score += when {
                rsi in 45.0..65.0  -> 20 // ideal entry zone
                rsi > 65 && rsi < 72 -> 10
                rsi in 30.0..<45.0 -> 14 // mild pullback
                rsi < 30           -> 16 // oversold — potential bounce
                else               -> 3  // overbought (>72) — avoid
            }

            score = minOf(100, score)

            val conviction = when {
                score >= 70 -> "high"
                score >= 50 -> "medium"
                score >= 30 -> "low"
                else        -> "noise"
            }

            if (conviction == "noise") continue

            val rsSign = if (sector.relativeStrength >= 0) "+" else ""
            val reasoning = listOf(
                "Parent: ${sector.label} (RS $rsSign${sector.relativeStrength}% vs SPY, ${sector.trend})",
                if (pressure.points.isNotEmpty())
                    "Pressure: ${pressure.points.joinToString("; ") { it.description }}"
                else "No significant pressure points",
                "RSI: ${pressure.rsi}",
            ).joinToString(" | ")

            signals.add(
                StockSignal(
                    symbol               = symbol,
                    label                = label,
                    parentSector         = sector.symbol,
                    parentSectorLabel    = sector.label,
                    parentRotationScore  = sector.score,
                    pressurePoints       = pressure.points,
                    rsi                  = rsi,
                    score                = score,
                    conviction           = conviction,
                    reasoning            = reasoning,
                )
            )
        }
    }

    return signals.sortedByDescending { it.score }
}
